src/plugins/action_generate_review_assignment_records/src/generateReviewAssignments.ts

```diff
@@ -1,112 +1,105 @@
+    const numReviewLevels: number =
+      (await DBConnect.getNumReviewLevels(templateId, stageNumber)) || 0
 
-    const currentReviewLevel: number = reviewId
-      ? await DBConnect.getCurrentReviewLevel(reviewId)
-      : 0
-    if (currentReviewLevel === numReviewLevels) {
-      console.log(
-        'Final review level reached for current stage, no later review assignments to generate.'
-      )
-      return {
-        status: 'Success',
-        error_log: '',
-        output: {},
+    // NB: reviewId comes from record_id on TriggerPayload when triggered from review table
+    // For first review assignment
+    if (!reviewId) {
+      console.log('First review assignment on first stage - total levels:', numReviewLevels)
+      if (numReviewLevels === 0) {
+        console.log(
+          'No reviewer with level associated to first stage, no review assignments to generate.'
+        )
+        return {
+          status: 'Success',
+          error_log: '',
+          output: {},
+        }
       }
+      const nextReviewLevel = 1
+      const isLastLevel = nextReviewLevel === numReviewLevels
+      return generateNextReviewAssignments({
+        db,
+        applicationId,
+        templateId,
+        nextReviewLevel,
+        nextStageNumber: stageNumber,
+        nextStageId: stageId,
+        isLastLevel,
+      })
     }
-    const nextReviewLevel = currentReviewLevel + 1
-
-    const nextLevelReviewers = await db.getPersonnelForApplicationStageLevel(
-      templateId,
-      stageNumber,
-      nextReviewLevel,
-      'Review'
-    )
-
-    const reviewAssignments: ReviewAssignmentObject = {}
-
-    // Build reviewers into object map so we can combine duplicate user_orgs
-    // and merge their section code restrictions
-    nextLevelReviewers.forEach((reviewer: Reviewer) => {
-      const { userId, orgId, restrictions } = reviewer
-
-      const templateSectionRestrictions = restrictions
-        ? restrictions?.templateSectionRestrictions
-        : null
-
-      const userOrgKey = `${userId}_${orgId ? orgId : 0}`
-      if (reviewAssignments[userOrgKey])
-        reviewAssignments[userOrgKey].templateSectionRestrictions = mergeSectionRestrictions(
-          reviewAssignments[userOrgKey].templateSectionRestrictions,
-          templateSectionRestrictions
-        )
-      else
-        reviewAssignments[userOrgKey] = {
-          reviewerId: userId,
-          orgId,
-          stageId,
-          stageNumber,
-          // TO-DO: allow STATUS to be configurable in template
-          status: nextReviewLevel === 1 ? AssignmentStatus.AVAILABLE : AssignmentStatus.SELF_ASSIGN,
+    // For level 1+ or next stages review assignment
+    // TODO - #440: Needs to lock same level reviews
+    // TODO: Unlock lower level review when consolidator has submitted "Change requested"
+    else {
+      const {
+        stageNumber: previousStage,
+        level: previousLevel,
+      } = await DBConnect.getReviewStageAndLevel(reviewId)
```

Comment:
These values should be available in applicationData, so no need to do another DB query. (That is, the applicationData that comes in on input, not a new one)